src/coprocessor/dag/expr/mod.rs

```diff
@@ -443,8 +447,42 @@ impl Expression {
     }
 }
 
+fn filter<T: Into<Datum>>(x: Result<T>) -> Option<Result<Datum>> {
+    match x {
+        Err(Error::UnknownSignature(_)) => None,
+        e => Some(e.map(|x| x.into())),
+    }
+}
+
 impl Expression {
-    fn build(mut expr: Expr, row_len: usize, ctx: &StatementContext) -> Result<Self> {
+    pub fn eval(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Datum> {
+        match *self {
+            Expression::Constant(ref constant) => Ok(constant.eval()),
+            Expression::ColumnRef(ref column) => Ok(column.eval(row)),
+            Expression::ScalarFn(ref f) => {
+                let eval_int = self.eval_int(ctx, row).map(|v| {
+                    v.map_or(Datum::Null, |v| {
+                        if mysql::has_unsigned_flag(self.get_tp().get_flag() as u64) {
+                            Datum::U64(v as u64)
+                        } else {
+                            Datum::I64(v)
+                        }
+                    })
+                });
+
+                let res = filter(eval_int)
+                    .or_else(|| filter(self.eval_real(ctx, row)))
```

Comment:
we match many times. Would it cause performance lose? @BusyJay

Reply:
I think this is only for test purpose.

Reply:
Let's merge it first to get things working first. I will figure out a way to reduce the lose.